File: kompartment/src/ode/julia/core/jacobian.js
/* ==========================================================================
   ode_julia / core / jacobian

   Getting df/du, and turning it into the matrix the solvers actually invert.

   Two things live here.

   1. JacobianCache -- produces J, either by calling a Jacobian the caller
      supplied or by differencing f. Differencing costs one evaluation of the
      whole right-hand side per *group* of structurally orthogonal columns, not
      per column, so a sparsity pattern is worth giving even when the matrix is
      then handled densely: a chemical Jacobian of 63 species needs about 45
      groups, and one of 200 species often still needs fewer than 50.

   2. WFactorization -- forms W = I − γh·J (or I/(γh) − J, which Radau and the
      Rosenbrock methods want) and factorises it, dense or sparse, and knows
      when it may skip doing so. That last part is where the time goes: a stiff
      solver that re-forms and re-factorises W on every step of a smooth
      stretch does several times the work of one that does not.
   ========================================================================== */

import {
  DenseMatrix, CSC, DenseLU, SparseLU, cscFromTriplets, reverseCuthillMcKee,
} from './linalg.js';

const SQRT_EPS = Math.sqrt(Number.EPSILON);   // ~1.49e-8

/**
 * Group the columns of a sparsity pattern so that no two columns in a group
 * share a row.
 *
 * Greedy distance-1 colouring of the column-intersection graph, largest degree
 * first. Perturbing every column of a group at once and differencing gives
 * each of their columns exactly, because no row receives a contribution from
 * more than one of them.
 *
 * @param {number} n
 * @param {Int32Array} colPtr
 * @param {Int32Array} rowIdx
 * @returns {{colour: Int32Array, groups: Int32Array[], count: number}}
 */
export function colourColumns(n, colPtr, rowIdx) {
  // rows -> the columns that touch them, so two columns can be seen to clash.
  const rowCount = new Int32Array(n + 1);
  for (let j = 0; j < n; j++) {
    for (let k = colPtr[j]; k < colPtr[j + 1]; k++) rowCount[rowIdx[k] + 1]++;
  }
  for (let i = 0; i < n; i++) rowCount[i + 1] += rowCount[i];
  const rowCols = new Int32Array(rowCount[n]);
  const at = Int32Array.from(rowCount);
  for (let j = 0; j < n; j++) {
    for (let k = colPtr[j]; k < colPtr[j + 1]; k++) rowCols[at[rowIdx[k]]++] = j;
  }

  const order = Array.from({ length: n }, (_, j) => j)
    .sort((a, b) => (colPtr[b + 1] - colPtr[b]) - (colPtr[a + 1] - colPtr[a]));

  const colour = new Int32Array(n).fill(-1);
  const taken = new Uint8Array(n + 1);
  let count = 0;
  const touched = [];
  for (const j of order) {
    touched.length = 0;
    for (let k = colPtr[j]; k < colPtr[j + 1]; k++) {
      const i = rowIdx[k];
      for (let p = rowCount[i]; p < rowCount[i + 1]; p++) {
        const other = rowCols[p];
        const c = colour[other];
        if (c >= 0 && !taken[c]) { taken[c] = 1; touched.push(c); }
      }
    }
    let c = 0;
    while (taken[c]) c++;
    colour[j] = c;
    if (c + 1 > count) count = c + 1;
    for (const t of touched) taken[t] = 0;
  }

  const sizes = new Int32Array(count);
  for (let j = 0; j < n; j++) sizes[colour[j]]++;
  const groups = Array.from({ length: count }, (_, c) => new Int32Array(sizes[c]));
  const fillAt = new Int32Array(count);
  for (let j = 0; j < n; j++) {
    const c = colour[j];
    groups[c][fillAt[c]++] = j;
  }
  return { colour, groups, count };
}

/** A dense n×n pattern, for when the caller gives none. */
export function densePattern(n) {
  const colPtr = new Int32Array(n + 1);
  const rowIdx = new Int32Array(n * n);
  for (let j = 0; j < n; j++) {
    colPtr[j + 1] = (j + 1) * n;
    for (let i = 0; i < n; i++) rowIdx[j * n + i] = i;
  }
  return { colPtr, rowIdx };
}

/**
 * Produces J = df/du at (t, u).
 *
 * Three ways in, in order of preference:
 *   jac(t, u, J)      the caller fills J themselves -- exact, and cheapest
 *   jacPattern + f    differenced, one f per colour group
 *   f alone           differenced, one f per column (n+1 evaluations a time)
 */
export class JacobianCache {
  /**
   * @param {number} n
   * @param {object} opts
   * @param {(t:number,u:Float64Array,J:object)=>void} [opts.jac]  fills J in place
   * @param {{colPtr:Int32Array,rowIdx:Int32Array}} [opts.jacPattern]
   * @param {'auto'|'sparse'|'dense'} [opts.matrix]
   * @param {boolean} [opts.central]  central differences (2 f per group, one more digit)
   */
  constructor(n, opts = {}) {
    this.n = n;
    this.userJac = opts.jac || null;
    this.central = !!opts.central;
    this.pattern = opts.jacPattern || null;
    this.sparse = false;
    this.nf = 0;         // f evaluations spent on Jacobians
    this.njac = 0;

    // Which storage J lives in. Sparse only pays when the pattern is given and
    // genuinely sparse; the factorisation decides separately, since a sparse J
    // can still have a factor so full that a dense LU is faster.
    const want = opts.matrix || 'auto';
    const nnz = this.pattern ? this.pattern.colPtr[n] : n * n;
    this.sparse = want === 'sparse' || (want === 'auto' && !!this.pattern && nnz < 0.25 * n * n);

    if (!this.pattern && (this.sparse || !this.userJac)) this.pattern = densePattern(n);

    if (this.sparse) {
      this.J = new CSC(n, this.pattern.colPtr, this.pattern.rowIdx);
    } else {
      this.J = DenseMatrix.zeros(n);
    }

    // Differencing machinery. Built even when a `jac` was given, because that
    // callback may answer `false` for a point it will not vouch for -- see
    // `evaluate` -- and differencing that one call is the whole answer to it.
    // A colouring at construction and three vectors is what it costs.
    this.groups = colourColumns(n, this.pattern.colPtr, this.pattern.rowIdx).groups;
    this.upert = new Float64Array(n);
    this.fpert = new Float64Array(n);
    this.fpert2 = this.central ? new Float64Array(n) : null;
    this.delta = new Float64Array(n);
  }

  /** How many right-hand-side evaluations one differenced Jacobian costs. */
  get groupCount() { return this.groups ? this.groups.length : 0; }

  /**
   * Fill this.J with df/du at (t, u). `fu` is f(t, u), already computed by the
   * caller -- forward differencing needs it and recomputing it is a waste.
   *
   * @param {(t:number,u:Float64Array,out:Float64Array)=>void} f
   */
  evaluate(f, t, u, fu) {
    this.njac++;
    // A caller's `jac` may answer `false` to mean "not this time" -- the
    // Ecolego port's analytic matrix declines a point where an entry comes out
    // non-finite, and differences for that one call rather than handing over a
    // matrix with an infinity in it. Anything else, including undefined, is
    // the matrix having been filled.
    if (this.userJac && this.userJac(t, u, this.J) !== false) return this.J;
    const { n, groups, upert, fpert, delta } = this;
    const { colPtr, rowIdx } = this.pattern;
    const values = this.sparse ? this.J.values : this.J.data;
    if (this.sparse) values.fill(0); else this.J.data.fill(0);

    for (const group of groups) {
      // One perturbation vector for the whole group.
      upert.set(u);
      for (let g = 0; g < group.length; g++) {
        const j = group[g];
        // The step the reference implementations use: relative to the component, but
        // never smaller than the same fraction of 1, so a component sitting at
        // zero still gets differenced.
        const d = SQRT_EPS * Math.max(Math.abs(u[j]), 1e-5);
        delta[j] = d;
        upert[j] = u[j] + d;
      }
      this.nf++;
      f(t, upert, fpert);
      if (this.central) {
        for (let g = 0; g < group.length; g++) {
          const j = group[g];
          upert[j] = u[j] - delta[j];
        }
        this.nf++;
        f(t, upert, this.fpert2);
      }

      for (let g = 0; g < group.length; g++) {
        const j = group[g];
        const scale = this.central ? 1 / (2 * delta[j]) : 1 / delta[j];
        for (let k = colPtr[j]; k < colPtr[j + 1]; k++) {
          const i = rowIdx[k];
          const back = this.central ? this.fpert2[i] : fu[i];
          const v = (fpert[i] - back) * scale;
          if (this.sparse) values[k] = v;
          else this.J.data[j * n + i] = v;
        }
      }
    }
    return this.J;
  }
}

/**
 * W = I − γh·J, or W = I/(γh) − J, formed and factorised.
 *
 * `transform` picks between them. The first is what a Newton iteration on a
 * DIRK or BDF stage wants; the second is what the Rosenbrock methods and Radau
 * want, because their stage equations are already written with the increments
 * divided through by γh, and forming W that way keeps the linear system's
 * scaling independent of the step size.
 *
 * The factorisation is kept and reused. `needsUpdate` is the whole economy of
 * a stiff solver: J is re-differenced only when the Newton iteration says it
 * has gone stale, and W is re-factorised only when J changed or γh moved
 * enough to matter.
 */
export class WFactorization {
  /**
   * @param {number} n
   * @param {JacobianCache} jacCache
   * @param {object} [opts]
   * @param {'auto'|'sparse'|'dense'} [opts.matrix]
   * @param {boolean} [opts.reorder]  apply a fill-reducing ordering (sparse only)
   */
  constructor(n, jacCache, opts = {}) {
    this.n = n;
    this.jacCache = jacCache;
    this.sparse = jacCache.sparse;
    this.nfactor = 0;
    this.nsolve = 0;
    this.fill = null;
    this.ordering = 'none';

    if (this.sparse) {
      // W has the pattern of J plus a full diagonal, which J may not carry.
      const J = jacCache.J;
      const rows = []; const cols = [];
      for (let j = 0; j < n; j++) {
        for (let k = J.colPtr[j]; k < J.colPtr[j + 1]; k++) { rows.push(J.rowIdx[k]); cols.push(j); }
        rows.push(j); cols.push(j);
      }
      this.W = cscFromTriplets(n, rows, cols);
      // Where each entry of J lands in W, worked out once.
      this.jToW = new Int32Array(J.colPtr[n]);
      for (let j = 0; j < n; j++) {
        for (let k = J.colPtr[j]; k < J.colPtr[j + 1]; k++) {
          const i = J.rowIdx[k];
          let p = this.W.colPtr[j];
          while (this.W.rowIdx[p] !== i) p++;
          this.jToW[k] = p;
        }
      }
      this.diagW = new Int32Array(n);
      for (let j = 0; j < n; j++) {
        let p = this.W.colPtr[j];
        while (this.W.rowIdx[p] !== j) p++;
        this.diagW[j] = p;
      }
      const perm = opts.reorder === false
        ? null
        : reverseCuthillMcKee(n, this.W.colPtr, this.W.rowIdx);
      this.ordering = perm ? 'reverse Cuthill-McKee' : 'natural';
      this.lu = new SparseLU(n, this.W.colPtr, this.W.rowIdx, perm);
    } else {
      this.W = DenseMatrix.zeros(n);
      this.lu = new DenseLU(n);
    }

    // What the stored factorisation was built from.
    this.gammaDt = NaN;
    this.transform = false;
    this.jacT = NaN;
    this.jacStale = true;
    this.haveFactor = false;

    // How many steps the stored Jacobian has been used for, and the most it
    // may be. A stiff solver saves most of its work by reusing J, and the
    // usual rule -- reuse while the Newton iteration keeps converging quickly
    // -- has a hole in it: a Newton seeded close enough to converge on its
    // first iteration always reports fast convergence, so J is never
    // questioned. TRBDF2 on the pollution problem walked off with a Jacobian
    // hundreds of steps old, converged happily at every step to the wrong
    // root, and reached 1e9 in components that belong between 0 and 1. By then
    // the relative error test scales with the blown-up state and notices
    // nothing. So there is also a plain age limit, as CVODE has (MSBP = 20).
    this.age = 0;
    this.maxAge = opts.maxJacAge ?? 20;
  }

  /** Say that J is out of date, so the next form() re-evaluates it. */
  markStale() { this.jacStale = true; }

  /** One accepted step has passed. */
  agePlus() { this.age++; }

  /** Has the stored Jacobian been used for as long as it may be? */
  get tooOld() { return this.age >= this.maxAge; }

  /** Is the stored J the one for this (t, u)? */
  isCurrent(t) { return !this.jacStale && this.jacT === t; }

  /**
   * Ensure J is current, then build and factorise W for this γh.
   *
   * @returns {boolean} false if W is singular, which the caller answers by
   *                    cutting the step rather than by giving up.
   */
  form(f, t, u, fu, gammaDt, transform, forceJac = false) {
    const n = this.n;
    // Only the caller decides. An earlier version also re-evaluated whenever t
    // had moved, which is every step, so the Jacobian reuse that the whole
    // design is built around never happened once: FBDF reported exactly as
    // many Jacobians as accepted steps.
    if (forceJac || this.jacStale || this.tooOld) {
      this.jacCache.evaluate(f, t, u, fu);
      this.jacT = t;
      this.jacStale = false;
      this.age = 0;
      this.haveFactor = false;
    } else if (this.haveFactor && this.gammaDt === gammaDt && this.transform === transform) {
      return true;             // nothing has changed: reuse the factor
    }

    const J = this.jacCache.J;
    if (this.sparse) {
      const w = this.W.values;
      w.fill(0);
      const jv = J.values;
      if (transform) {
        const d = 1 / gammaDt;
        for (let k = 0; k < jv.length; k++) w[this.jToW[k]] = -jv[k];
        for (let j = 0; j < n; j++) w[this.diagW[j]] += d;
      } else {
        for (let k = 0; k < jv.length; k++) w[this.jToW[k]] = -gammaDt * jv[k];
        for (let j = 0; j < n; j++) w[this.diagW[j]] += 1;
      }
      this.nfactor++;
      const ok = this.lu.factor(this.W.colPtr, this.W.rowIdx, w);
      this.fill = this.lu.fill;
      this.haveFactor = ok;
      this.gammaDt = gammaDt;
      this.transform = transform;
      return ok;
    }

    const w = this.W.data;
    const jd = J.data;
    if (transform) {
      const d = 1 / gammaDt;
      for (let k = 0; k < w.length; k++) w[k] = -jd[k];
      for (let j = 0; j < n; j++) w[j * n + j] += d;
    } else {
      for (let k = 0; k < w.length; k++) w[k] = -gammaDt * jd[k];
      for (let j = 0; j < n; j++) w[j * n + j] += 1;
    }
    this.nfactor++;
    const ok = this.lu.factor(this.W);
    this.haveFactor = ok;
    this.gammaDt = gammaDt;
    this.transform = transform;
    return ok;
  }

  /** Solve Wx = b in place. */
  solve(b) {
    this.nsolve++;
    this.lu.solve(b);
    return b;
  }
}

File: kompartment/python/kompartment/engine/solvers/julia/jacobian.py
"""df/du, and the matrix the solvers invert (``core/jacobian.js``).

``JacobianCache`` produces J -- from the caller's Jacobian, which may decline
a point, or by forward differences through the package's own colouring of the
pattern -- and ``WFactorization`` forms W = I - gh*J (or I/(gh) - J) from it,
factorises it, and knows when it may reuse the factor or must re-evaluate J.
"""

from __future__ import annotations

import math
from typing import Callable, List, Optional, Tuple

import numpy as np

from ._js import SQRT_EPS
from .linalg import DenseLU, SparseLU, csc_from_triplets, dense_pattern, reverse_cuthill_mckee

Pattern = Tuple[np.ndarray, np.ndarray]


def colour_columns(n: int, col_ptr: np.ndarray, row_idx: np.ndarray) -> List[np.ndarray]:
    """Groups of columns that share no row: greedy, largest degree first, a
    stable sort (``colourColumns``). Each group lists its columns ascending."""
    cp = np.asarray(col_ptr, dtype=np.int64)
    ri = np.asarray(row_idx, dtype=np.int64)
    # rows -> the columns that touch them
    row_count = np.zeros(n + 1, dtype=np.int64)
    np.cumsum(np.bincount(ri, minlength=n), out=row_count[1:])
    col_of = np.repeat(np.arange(n, dtype=np.int64), np.diff(cp))
    by_row = np.argsort(ri, kind='stable')
    row_cols = col_of[by_row].tolist()
    rc = row_count.tolist()
    cpl = cp.tolist()
    ril = ri.tolist()
    degree = np.diff(cp).tolist()
    order = sorted(range(n), key=lambda j: -degree[j])
    colour = [-1] * n
    taken = [False] * (n + 1)
    count = 0
    for j in order:
        touched = []
        for k in range(cpl[j], cpl[j + 1]):
            i = ril[k]
            for p in range(rc[i], rc[i + 1]):
                c = colour[row_cols[p]]
                if c >= 0 and not taken[c]:
                    taken[c] = True
                    touched.append(c)
        c = 0
        while taken[c]:
            c += 1
        colour[j] = c
        if c + 1 > count:
            count = c + 1
        for t in touched:
            taken[t] = False
    colours = np.array(colour, dtype=np.int64)
    by_colour = np.argsort(colours, kind='stable')
    sizes = np.bincount(colours, minlength=count) if n else np.zeros(0, dtype=np.int64)
    return np.split(by_colour, np.cumsum(sizes)[:-1]) if count else []


class JacobianCache:
    """J = df/du at (t, u) (``JacobianCache``).

    ``user_jac(t, u, cache)`` fills ``cache.values`` (sparse) or ``cache.J``
    (dense) and answers False to have this one point differenced instead.
    """

    def __init__(self, n: int, user_jac: Optional[Callable[..., bool]], jac_pattern: Optional[Pattern],
                 matrix: Optional[str], central: bool) -> None:
        self.n = n
        self.user_jac = user_jac
        self.central = bool(central)
        self.pattern = jac_pattern
        self.nf = 0
        self.njac = 0
        want = matrix or 'auto'
        nnz = int(jac_pattern[0][n]) if jac_pattern is not None else n * n
        self.sparse = want in ('sparse', 'refactor') or (want == 'auto' and jac_pattern is not None
                                                          and nnz < 0.25 * n * n)
        if self.pattern is None and (self.sparse or user_jac is None):
            self.pattern = dense_pattern(n)
        if self.sparse:
            self.values = np.zeros(int(self.pattern[0][n]))
            self.J: Optional[np.ndarray] = None
        else:
            self.values = None  # type: ignore[assignment]
            self.J = np.zeros((n, n))
        self.diff_pattern = self.pattern
        self.groups: Optional[List[np.ndarray]] = None
        self._plan: Optional[list] = None
        if self.pattern is not None:
            self._set_groups(colour_columns(n, self.pattern[0], self.pattern[1]))

    def _set_groups(self, groups: List[np.ndarray]) -> None:
        """The colouring, and per group the pattern entries it fills."""
        self.groups = groups
        col_ptr, row_idx = self.diff_pattern  # type: ignore[misc]
        plan = []
        lens = np.diff(col_ptr)
        for g in groups:
            glens = lens[g]
            total = int(glens.sum())
            starts = np.asarray(col_ptr, dtype=np.int64)[g]
            first = np.zeros(g.size, dtype=np.int64)
            if g.size:
                np.cumsum(glens[:-1], out=first[1:])
            entries = np.repeat(starts, glens) + (np.arange(total, dtype=np.int64) - np.repeat(first, glens))
            which = np.repeat(np.arange(g.size, dtype=np.int64), glens)
            plan.append((g, entries, np.asarray(row_idx, dtype=np.int64)[entries], which))
        self._plan = plan

    @property
    def group_count(self) -> int:
        return len(self.groups) if self.groups is not None else 0

    def evaluate(self, f: Callable[[float, np.ndarray], np.ndarray], t: float, u: np.ndarray, fu: np.ndarray) -> None:
        self.njac += 1
        if self.user_jac is not None and self.user_jac(t, u, self) is not False:
            return
        if self.groups is None:
            self.diff_pattern = dense_pattern(self.n)
            self._set_groups(colour_columns(self.n, self.diff_pattern[0], self.diff_pattern[1]))
        n = self.n
        if self.sparse:
            vals = self.values
            vals.fill(0.0)
        else:
            self.J.fill(0.0)  # type: ignore[union-attr]
            vals = None
        for g, entries, rows, which in self._plan:  # type: ignore[union-attr]
            upert = u.copy()
            ug = u[g]
            d = SQRT_EPS * np.maximum(np.abs(ug), 1e-5)
            upert[g] = ug + d
            self.nf += 1
            fpert = f(t, upert)
            if self.central:
                upert[g] = ug - d
                self.nf += 1
                back = f(t, upert)
                scale = 1.0 / (2.0 * d)
            else:
                back = fu
                scale = 1.0 / d
            v = (fpert[rows] - back[rows]) * scale[which]
            if vals is not None:
                vals[entries] = v
            else:
                self.J[rows, g[which]] = v  # type: ignore[index]


class WFactorization:
    """W = I - gh*J, or I/(gh) - J with ``transform``, formed and factorised
    (``WFactorization``)."""

    def __init__(self, n: int, jac_cache: JacobianCache, max_jac_age: Optional[float] = None,
                 reorder: bool = True) -> None:
        self.n = n
        self.jac_cache = jac_cache
        self.sparse = jac_cache.sparse
        self.nfactor = 0
        self.nsolve = 0
        # Whether a sparse factorisation has been attempted: the package's
        # `fill` is null until then, and the latest successful factor's after.
        self._factored = False
        self.ordering = 'none'
        if self.sparse:
            j_ptr, j_row = jac_cache.pattern  # type: ignore[misc]
            j_ptr = np.asarray(j_ptr, dtype=np.int64)
            j_row = np.asarray(j_row, dtype=np.int64)
            j_col = np.repeat(np.arange(n, dtype=np.int64), np.diff(j_ptr))
            diag = np.arange(n, dtype=np.int64)
            w_ptr, w_row = csc_from_triplets(n, np.concatenate([j_row, diag]), np.concatenate([j_col, diag]))
            self.w_ptr = w_ptr
            self.w_row = w_row
            w_col = np.repeat(np.arange(n, dtype=np.int64), np.diff(w_ptr))
            keys = w_col * n + w_row
            self.j_to_w = np.searchsorted(keys, j_col * n + j_row)
            self.diag_w = np.searchsorted(keys, diag * n + diag)
            self.w_values = np.zeros(w_row.size)
            perm = None if reorder is False else reverse_cuthill_mckee(n, w_ptr, w_row)
            self.ordering = 'reverse Cuthill-McKee' if perm is not None else 'natural'
            self.lu = SparseLU(n, w_ptr, w_row, perm)
        else:
            self.W = np.zeros((n, n))
            self._diag = np.arange(n)
            self.lu = DenseLU(n)  # type: ignore[assignment]
        self.gamma_dt = math.nan
        self.transform = False
        self.jac_t = math.nan
        self.jac_stale = True
        self.have_factor = False
        self.age = 0
        self.max_age = max_jac_age if max_jac_age is not None else 20

    def mark_stale(self) -> None:
        self.jac_stale = True

    def age_plus(self) -> None:
        self.age += 1

    @property
    def too_old(self) -> bool:
        return self.age >= self.max_age

    def is_current(self, t: float) -> bool:
        return not self.jac_stale and self.jac_t == t

    def form(self, f: Callable[[float, np.ndarray], np.ndarray], t: float, u: np.ndarray, fu: np.ndarray,
             gamma_dt: float, transform: bool, force_jac: bool = False) -> bool:
        if force_jac or self.jac_stale or self.too_old:
            self.jac_cache.evaluate(f, t, u, fu)
            self.jac_t = t
            self.jac_stale = False
            self.age = 0
            self.have_factor = False
        elif self.have_factor and self.gamma_dt == gamma_dt and self.transform == transform:
            return True
        jc = self.jac_cache
        if self.sparse:
            w = self.w_values
            w.fill(0.0)
            if transform:
                d = 1 / gamma_dt
                w[self.j_to_w] = -jc.values
                w[self.diag_w] += d
            else:
                w[self.j_to_w] = -gamma_dt * jc.values
                w[self.diag_w] += 1
            self.nfactor += 1
            ok = self.lu.factor(w)  # type: ignore[call-arg]
            self._factored = True
            self.have_factor = ok
            self.gamma_dt = gamma_dt
            self.transform = transform
            return ok
        W = self.W
        if transform:
            d = 1 / gamma_dt
            np.negative(jc.J, out=W)
            W[self._diag, self._diag] += d
        else:
            np.multiply(jc.J, -gamma_dt, out=W)
            W[self._diag, self._diag] += 1
        self.nfactor += 1
        ok = self.lu.factor(W)
        self.have_factor = ok
        self.gamma_dt = gamma_dt
        self.transform = transform
        return ok

    @property
    def fill(self) -> Optional[int]:
        """The package's ``fill``: null for a dense W or before the first
        factorisation, else the latest successful sparse factor's."""
        if not self.sparse or not self._factored:
            return None
        return self.lu.fill  # type: ignore[union-attr]

    def solve(self, b: np.ndarray) -> np.ndarray:
        self.nsolve += 1
        return self.lu.solve(b)
